text/0000-macro-expansion-for-macro-input.md

```diff
@@ -0,0 +1,214 @@
+
+* Leads to frustrating corner-cases involving macro paths. For instance, consider the following:
+
+    ```rust
+    macro baz!(...);
+    foo! {
+        mod b {
+            super::baz!();
+        }
+    }
+    ```
+
+    The caller of `foo!` probably imagines that `baz!` will be expanded within `mod b`, and so prepends the call with `super`. However, if `foo!` naively marks the call to `super::baz!`, then the path will fail to resolve because macro paths are resolved relative to the location of the call. Handling this would require the macro implementer to track the path offset of its expansion, which is doable but adds complexity.
+
+* Commits the compiler to a particular (but loose) macro expansion order, as well as a (limited) way for users to position themselves within that order. What future plans does this interfere with? What potentially unintuitive expansion-order effects might this expose?
+
+# Rationale and alternatives
+
+The primary rationale is to make procedural and attribute macros work more smoothly with other features of Rust - mainly other macros.
+
+Recalling the examples listed in the [motivation](#motivation) above, a few but not all situations of proc macros receiving unexpanded macro calls could be avoided by changing the general 'hands off' attitude towards proc macros and attribute macros, and more aggressively parse and expand their inputs. This effectively bans macro calls as part of the input grammar, which seems drastic, and wouldn't handle cases of indirection via token tree (`$x:tt`) parameters.
+
+We could encourage the creation of a 'macros for macro authors' crate with implementations of common macros - for instance, those in the standard library - and make it clear that macro support isn't guaranteed for arbitrary macro calls passed in to proc macros. This feels unsatisfying, since it fractures the macro ecosystem and leads to very indirect unexpected behaviour (for instance, one proc macro may use a different macro expansion library than another, and they might return different results). This also doesn't help address macro calls in built-in attributes.
+
+# Unresolved questions
+
+* This API allows for a first-pass solution to the problems listed in the [motivation](#motivation). Does it interfere with any known uses of proc macros? Does it prevent any existing techniques from working or cut off potential future ones?
+
+* How does this proposal affect expansion within the _body_ of an attribute macro call? Currently builtin macros like `#[cfg]` are special-cased to expand before things like `#[derive]`; can we unify this behaviour under the new system?
+
+* How does this handle structured arguments passed to declarative macros (like `$x:expr`)?
```

Comment:
This question needs some clarification: Do you mean that the argument to a procedural macro might contain a macro declaration?